### PR TITLE
Feature/13

### LINKED ISSUE #13: Add Canvas Showcase
3D Canvas
VR Canvas

### PR DESCRIPTION
**Pull request by afonsobspinto**
Closes #13 

Have a look at https://github.com/openworm/geppetto-client/pull/332 for details.

Does not include the flip feature which was staling the PR.

When merged the following issues should be created:
- Add flip feature to Canvas
- Handle question:
> If, on the application, I do not setup the color I want for an instance that I provide to the Canvas (proxyinstance) and the Canvas automatically sets up a color for it, how does the application knows which color was assigned?
- Refactor the canvas not to use capabilities, better look for the position. from @filippomc 

The aforementioned issues are not handled in this PR due to not affecting the standard behavior of the canvas which besides those details seems to be ready

### CODE REVIEW

File: geppetto-showcase/package.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "@babel/plugin-proposal-class-properties": "^7.8.3",
     "@babel/preset-env": "^7.8.4",
     "@babel/preset-react": "^7.8.3",
-    "@geppettoengine/geppetto-client": "file:../geppetto.js",
+    "@geppettoengine/geppetto-client": "file:.yalc/@geppettoengine/geppetto-client",
```
Comment (lrebscher) [inline geppetto-showcase/package.json:17]: there is an open PR #53  for this feature, this explains why the deployment here failed, yalc steps need to be added to the Dockerfile as well

File: geppetto.js/geppetto-ui/src/3d-canvas/Canvas.js [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+      cameraOptions,
+      cameraHandler,
+      selectionHandler,
+      backgroundColor,
+      pickingEnabled,
+      linesThreshold,
+      hoverListeners,
+      onMount
+    } = this.props;
+    // TODO: pass props instead?
```
Comment (lrebscher) [inline geppetto.js/geppetto-ui/src/3d-canvas/Canvas.js:35]: I would go for explicit here and leave it as it is

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/VFBExample.js [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+          instancePath: 'VFB_00030783',
+          color: COLORS[6],
+        },
+      ],
+      selected: {},
+      threeDObjects: [],
+      modelVersion:0
+    };
+
+    this.cameraOptions = {
```
Comment (lrebscher) [inline geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/VFBExample.js:82]: in the example before cameraOptions was part of the state , why not here?

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        },
+        reset: false,
+        autorotate: false,
+        wireframe: false,
+        position: { x: 230.357, y: 256.435, z: 934.238 },
+        rotation: { rx: -0.294, ry: -0.117, rz: -0.02, radius: 531.19 },
+      },
+    };
+
+    this.lastCameraUpdate = null;
```
Comment (lrebscher) [inline geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js:66]: don't understand why we keep track of lastCameraUpdate here in this way, just curious, no change request here

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+      document.addEventListener('mousedown', this.handleClickOutside);
+    })
+  }
+  
+  componentWillUnmount () {
+    document.removeEventListener('mousedown', this.handleClickOutside);
+  }
+  
+  handleClickOutside (event) {
+
```
Comment (lrebscher) [inline geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js:89]: remove empty line

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+      Instances.getInstance(INSTANCE_NAME);
+      this.setState({ hasModelLoaded: true, showLoader: false })
+    })
+  }
+
+  cameraHandler (obj) {
+    this.lastCameraUpdate = obj;
+  }
+
+  selectionHandler (selectedMap) {
```
Comment (lrebscher) [inline geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js:111]: we should generalize this at some point and provide a default behavior and think about how to simplify custom handlers

cannot expect devs to have such in depth knowledge about the Canvas :D

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample2.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+const SELECTION_COLOR = { r: 0.8, g: 0.8, b: 0, a: 1 };
+
+const styles = () => ({
+  container: {
+    height: '800px',
+    width: '1240px',
+    display: 'flex',
+    alignItems: 'stretch',
+  },
+});
```
Comment (lrebscher) [inline geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample2.js:24]: there should be an empty line between class definitions, need to update the eslint config at some point (#63)

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample2.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+    this.lastCameraUpdate = null;
+    this.cameraHandler = this.cameraHandler.bind(this);
+    this.selectionHandler = this.selectionHandler.bind(this);
+    this.hoverHandler = this.hoverHandler.bind(this);
+    this.handleClickOutside = this.handleClickOutside.bind(this);
+    this.handleToggle = this.handleToggle.bind(this);
+  }
+
+  componentDidMount () {
```
Comment (lrebscher) [inline geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample2.js:84]: empty lines between methods

**Review by lrebscher — CHANGES_REQUESTED** [review]
Nice job! 👍

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js (hunk 4 above)
Comment (afonsobspinto) [reply geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js:66]: We (now) control the camera (position, rotation) through props (camOptions).
However, the movement in the scene doesn't go through the canvas component but rather affects the scene directly (using Threejs TrackballControls).
If we don't keep track of the cameraUpdates (using the camera handler) future renders of the canvas component will pass a past position in the camOptions props (making the camera jump to a previous place)

Comment (afonsobspinto) [reply geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js:66]: Added this issue to #67

File: geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js (hunk 6 above)
Comment (lrebscher) [reply geppetto.js/geppetto-ui/src/3d-canvas/showcase/examples/AuditoryCortexExample.js:110]: 👍

**Pull request merged**